Source/WebCore/dom/Document.cpp

```diff
@@ -9019,20 +9025,51 @@ HTMLElement* Document::topmostAutoPopover() const
 // https://html.spec.whatwg.org/#hide-all-popovers-until
 void Document::hideAllPopoversUntil(Element* endpoint, FocusPreviousElement focusPreviousElement, FireEvents fireEvents)
 {
-    Vector<Ref<HTMLElement>> popoversToHide;
-    for (auto& item : makeReversedRange(m_topLayerElements)) {
-        if (!is<HTMLElement>(item))
-            continue;
-        auto& element = downcast<HTMLElement>(item.get());
-        if (element.popoverState() == PopoverState::Auto) {
-            if (&element == endpoint)
-                break;
-            popoversToHide.append(element);
+    auto closeAllOpenPopovers = [&]() {
+        while (RefPtr popover = topmostAutoPopover())
+            popover->hidePopoverInternal(focusPreviousElement, fireEvents);
+    };
+    if (!endpoint)
+        return closeAllOpenPopovers();
+    auto autoPopoverList = [&]()  {
+        Vector<Ref<HTMLElement>> popovers;
+        for (auto& item : m_topLayerElements) {
+#if ENABLE(FULLSCREEN_API)
+            if (item->hasFullscreenFlag())
+                continue;
+#endif
+            if (auto* candidate = dynamicDowncast<HTMLElement>(item.get()); candidate && candidate->popoverState() == PopoverState::Auto) {
+                if (!is<HTMLDialogElement>(candidate) || !downcast<HTMLDialogElement>(candidate)->isOpen())
+                    popovers.append(*candidate);
+            }
         }
-    }
+        return popovers;
+    };
 
-    for (auto& popover : popoversToHide)
-        popover->hidePopoverInternal(focusPreviousElement, fireEvents);
+    bool repeatingHide = false;
+    do {
+        bool foundEndPoint = false;
+        RefPtr<Element> lastToHide;
```

Comment:
nit: Can we swap the two lines to match the spec?

Reply:
Done.